Added XDP support to BCC

Eichhoernchen opened the pull request:
This PR adds XDP support as currently supported in net-next.

Concretely, it adds two functions to bcc, namely: 
`attach_xdp` and `remove_xdp`
which allows to attach an XDP program to a device (given via its name, e.g., en0) (in the future this might change to a specific queue on a device once the kernel offers this interface)
and `remove_xdp` removes a XDP program from a device. Please note that there can currently be only one program attached to the device and attaching another program replaces the previous.

One example is available to test XDP, in networking/xdp which drops all packets an counts for which protocol a packet was dropped (this is taken from the kernel xdp1 example). Please note that you cannot use the network headers defined in `<bcc/proto.h>` as they cause llvm/clang to generate instructions not available on XDP layer. On XDP layer you do not have an skb yet, so you are operating on the bare packet data.


XDP support is currently limited to only some network adapters, there is the `mlx4` and there is also a patch available for the `e1000` driver.

4ast commented:
awesome. Thanks! lgtm @drzaeus77

drzaeus77 commented on examples/networking/xdp/xdp_drop_count.py:
```diff
@@ -0,0 +1,123 @@
+from bcc import BPF
```
I think it would help to chmod +x the flie that is checked in, as well as add `#/usr/bin/env python` as the first line.

Eichhoernchen replied:
Agree will fix!

drzaeus77 commented on examples/networking/xdp/xdp_drop_count.py:
```diff
@@ -0,0 +1,123 @@
+import time
+import sys
+
+
```
Can you please remove the whitespace?

Also, it would be good to add your own copyright notice to the top of the file.

Eichhoernchen replied:
Okay, can do

drzaeus77 commented on examples/networking/xdp/xdp_drop_count.py:
```diff
@@ -0,0 +1,123 @@
+
+# load BPF program
+b = BPF(text = """
+    #include <uapi/linux/bpf.h>
```
We've been taking the convention of aligning the embedded strings starting from column 0. That is, treat the string as if it were a C file starting from scratch, with indentation rules for C in effect after the opening """.

Eichhoernchen replied:
Okay, I will fix this

drzaeus77 commented on examples/networking/xdp/xdp_drop_count.py:
```diff
@@ -0,0 +1,123 @@
+# load BPF program
+b = BPF(text = """
+    #include <uapi/linux/bpf.h>
+    #include <linux/in.h>
```
When I ran this, I had to add:
```
#define KBUILD_MODNAME "foo"
```
In order to silence some warnings.

Eichhoernchen replied:
I also get a warning and I could not figure out the reason for it, can you give some insights why this silences the warning?

drzaeus77 commented on examples/networking/xdp/xdp_drop_count.py:
```diff
@@ -0,0 +1,123 @@
+
+
+
+    int xdp_prog1(struct xdp_md *ctx) {
```
Opening braces here are inconsistent. In bcc, we're mostly using same-line opening braces, meaning this line is correct and the `parse_ipv[46]` functions should be compacted by 1 line.

Eichhoernchen replied:
Will do!

drzaeus77 commented on src/cc/libbpf.c:
```diff
@@ -408,3 +408,104 @@ void * bpf_open_perf_buffer(perf_reader_raw_cb raw_cb, void *cb_cookie, int pid,
+                if (!err->error)
+                    continue;
+                fprintf(stderr, "bpf: nlmsg error %s\n", strerror(-err->error));
+		errno = -err->error;
```
Some hard tabs got mixed in here, please remove them.

Eichhoernchen replied:
Will do!

drzaeus77 commented on src/cc/libbpf.c:
```diff
@@ -408,3 +408,104 @@ void * bpf_open_perf_buffer(perf_reader_raw_cb raw_cb, void *cb_cookie, int pid,
+        if (nh->nlmsg_seq != seq) {
+            fprintf(stderr, "bpf: Wrong seq %d, expected %d\n",
+                   nh->nlmsg_seq, seq);
+	    errno = ESPIPE;
```
These seem like odd choices to me, and besides there is already a printf with a better error message. Might make more sense just to leave the printf and remove the os.strerror from the python side, and then leave errno alone.

Eichhoernchen replied:
Hm, I disagree. The c-printf only appears to the user when debug=4, otherwise noting meaningful would be printed. I chose these errno codes as they they came close to what happens there.

Eichhoernchen commented:
I will incorporate Brendens comments and update the PR tomorrow!

Eichhoernchen commented:
Hi,
I incorporated @drzaeus77 remarks and I also changed the error behavior.
Now errno is set to EBADMSG and the python part checks that message and recommends increasing the debug level

drzaeus77 commented on examples/networking/xdp/xdp_drop_count.py:
```diff
@@ -0,0 +1,135 @@
+# C code adapted from Brenden Blanco as found in kernel:samples/bpf/xdp1_kern.c
+#
+# Copyright (c) 2016 Jan Rüth
+# Licensed under the Apache License, Version 2.0 (the "License")
```
Hmm, I don't think you are allowed to do this (change license). Please drop this file and I will resubmit with the changed license.

Eichhoernchen replied:
Hm, yes I was unsure about it, as I only adapted parts in the C file, I think I have to adapt the GPL license, however the C file is actually not really part of the python file. I thought it would be ok to include the GPL note in the C part and just use the apache license (as it is used also in the other files) for the python part. would it be better to have them as separate files? Would that solve things? Do you mean drop the whole python file?
It would also be fine with me to just have the GPL license in the file, I don't know if the project requires anything.

drzaeus77 replied:
1. I would prefer to have this file checked in as ASL2
2. The original file is GPL (C) PLUMgrid, which you can include as-is but not re-license
3. I can re-license the original file as ASL2, on behalf of PLUMgrid (my employer)
4. Therefore, to achieve 1, I would prefer to be the author of this file, and will include a more appropriate license for the bcc project.
5. You can then later adapt this file, and add your (C), since it will be ASL2 already.

drzaeus77 replied:
In other words, yes please drop the whole python file, the end result will be cleaner from a licensing point of view.

Eichhoernchen replied:
ok great, thank you

drzaeus77 commented:
Also, if you would, please squash the commits. You can do it locally and just force push to your same branch, and github will pick up the new git log automatically, while still preserving these comments.

Eichhoernchen commented:
Okay, I removed the example and squashed the commits together.
Thanks for your patience and for making me aware how to use git properly.
Also thanks for your code review and your comments. This is my first push request, I hope I won't make the same mistakes the next time.

If there is still anything you like changed let me know

drzaeus77 commented:
Nicely done!
